File: packages/native/android/src/main/java/com/margelo/nitro/one/HybridOneCrypto.kt
package com.margelo.nitro.one

import com.margelo.nitro.core.ArrayBuffer
import java.security.SecureRandom

// secure random bytes for the crypto polyfill, from SecureRandom. an invalid
// count throws so the js side never falls back to Math.random.
class HybridOneCrypto : HybridOneCryptoSpec() {
    private val secureRandom = SecureRandom()

    override fun getRandomBytes(count: Double): ArrayBuffer {
        if (count.isNaN() || count < 0 || count > MAX_BYTES || count != kotlin.math.floor(count)) {
            throw IllegalArgumentException("secure random: invalid byte count $count")
        }
        val bytes = ByteArray(count.toInt())
        secureRandom.nextBytes(bytes)
        return ArrayBuffer.copy(bytes)
    }

    companion object {
        const val MAX_BYTES = 65536
    }
}

File: packages/one/ios/Nitro/HybridOneCrypto.swift
import NitroModules
import Security

// secure random bytes for the crypto polyfill, straight from
// SecRandomCopyBytes into the returned buffer. a failed generator throws so
// the js side never falls back to Math.random.
final class HybridOneCrypto: HybridOneCryptoSpec {
  func getRandomBytes(count: Double) throws -> ArrayBuffer {
    guard count >= 0, count <= 65536, count == count.rounded() else {
      throw RuntimeError.error(withMessage: "secure random: invalid byte count \(count)")
    }
    let length = Int(count)
    let buffer = ArrayBuffer.allocate(size: length)
    guard length == 0 || SecRandomCopyBytes(kSecRandomDefault, length, buffer.data) == errSecSuccess
    else {
      throw RuntimeError.error(withMessage: "secure random: SecRandomCopyBytes failed")
    }
    return buffer
  }
}
